### ISSUE TITLE
feature request: API-based custom sign-up and sign-in.

### ISSUE THREAD

**Issue by zuohuadong** [logto-io/logto#3787]
### What problem did you meet?
I have a program that wants to do automatic login without going through a login page, but there is no API to provide this.

It does not implement auth 2.0 and does not provide an interface.

### Describe what you'd like Logto to have

![image](https://user-images.githubusercontent.com/11203929/235046604-230c664f-583a-4395-a412-44318f038ead.png)

I can add users through it, but I can't sign in with a user.

**Comment by gao-sun** [comment]
Do Machine-to-Machine apps solve your issue? Or you want to programmatically fetch tokens that represents a user?

**Comment by zuohuadong** [comment]
> Do Machine-to-Machine apps solve your issue? Or you want to programmatically fetch tokens that represents a user?

No, because you can't get the username and password, and you can't customize the login to get the token

need:

https://next-auth.js.org/getting-started/rest-api

**Comment by gao-sun** [comment]
@wangsijie @simeng-li can you guys validate the request? thanks

**Comment by zuohuadong** [comment]
@gao-sun It is not a fully public API that can be tokenized or restricted by source IP. It can even be left to the developers themselves.

**Comment by September999999999** [comment]
@gao-sun First of all, my English is not very good, this translation is Chinese to English, I also encountered a similar problem, the part of our user login can not use oauth authorized page jump to complete, can only pass say api username password form of login, but in the logto did not find the api provided

**Comment by simeng-li** [comment]
@zuohuadong do you mind providing some more details of your use case?  As for a general to C application, an automatic sign-in flow without establishing a client session is prohibited.  
What kind of program do you have? Is it for internal development use only or for public usage?
Anyone who holds a valid username credential can get authorized on behave of that user?

**Comment by zuohuadong** [comment]
@simeng-li  wechat miniprogram  and  nodejs client.

**Comment by simeng-li** [comment]
So you would like to host a custom miniapp provided sigh-in form, simply call Logto to auth APIs through your miniapp

**Comment by zuohuadong** [comment]
@simeng-li If I could provide an API, I would be able to customize the provider

**Comment by Dav-id** [comment]
To add to this discussion, it would be valuable for me to be able to authenticate users via API by passing their username + password to get a JWT back. 

For example, I would do this using a custom UI inside of a Unity3D application where I would not load a webview due to the overhead of embedding chromium or similar inside of the application.

**Comment by wangsijie** [comment]
@Dav-id I believe this is a meaningful scenario. I will post updates here, so stay tuned.

**Comment by sskyy** [comment]
It is also valuable for E2E testing.

**Comment by JackyTYang** [comment]
![image](https://github.com/logto-io/logto/assets/56026984/d29ed8d8-2bc9-4d38-adaa-d9562c90de15)
I wonder what changed in the database when I sign up the first time l open logto:3002 in my browser, because I want implement the sign-up action just by manipulating the database and use the token acquired from database to use api to interact with logto.

**Comment by wangsijie** [comment]
@JackyTYang To interact with Logto management API, a machine-to-machine application is recommended. Would it be helpful if Logto has a feature to auto generate first machine to machine application after seeding?

**Comment by JackyTYang** [comment]
Sure, that would be very helpful. We’d appreciate it if you add that feature.

**Comment by wangsijie** [comment]
We have this feature on our roadmap but haven't locked in a specific timeline for its release. If it's something you're eager to see soon, we encourage contributions and would be thrilled to review a pull request from you.

**Comment by udleinati** [comment]
Hello, we have the same need here. We have a specific situation where we'd like to implement code-based login to obtain access to JWT. Similar to what can be done with Keycloak or Google Firebase. This feature would be greatly appreciated.

**Comment by johngrimsey** [comment]
Thanks for revisiting, +1 here for this feature

**Comment by systemis** [comment]
I have same issue, would love to have this feature.

**Comment by wangsijie** [comment]
We are refactoring the sign in experience APIs

**Comment by phpmaple** [comment]
Any progress? Do you have a plan to release it at a certain time? I really need this feature.

**Comment by Loongphy** [comment]
> Any progress? Do you have a plan to release it at a certain time? I really need this feature.@wangsijie

I think they're working on it. Not sure when it will be released. See [roadmap](https://logto.productlane.com/roadmap)  here.

**Comment by wangsijie** [comment]
The "Experience API" is currently in testing phase. It's expected to be ready for deployment soon. @simeng-li is leading this project. Please note that the new API still requires redirection and cookies to function properly.

**Comment by phpmaple** [comment]
> The "Experience API" is currently in testing phase. It's expected to be ready for deployment soon. @simeng-li is leading this project. Please note that the new API still requires redirection and cookies to function properly.

So, can I use it in the native app?  Redirection may not be useful in native apps.

**Comment by wangsijie** [comment]
> So, can I use it in the native app? Redirection may not be useful in native apps.

Cookies are required, you can implement it by in-app browser.

**Comment by wangsijie** [comment]
Then you can maintain the cookies by yourself, save "set-cookies" header and set it to headers of each request.
